Datetime Widget Config: "Protect" field format change possiblities

signedav opened the pull request:
User often confuses field format with display widget format. Mostly the user does not want to change the field format but thinks it needs to be aligned to the widget display format. After making there wrong settings, filtering does not work anymore etc.

### Changes:

![image](https://user-images.githubusercontent.com/28384354/222679257-8fc11a55-6003-4b8e-8349-691d4667e941.png)

- "Field Format" settings are now below "Display Format".
- "Widget Display" is now called "Display Format" (IMO Widget Display does not make much sense for users).
- "Field Format" is now in a checkable group box "Overwrite Field Format". User needs to activate it, to overwrite the default value. 
- When "Overwrite Field Format" gets unchecked again, the Field Format Combobox resets back to "Default" (this to avoid that user could think the custom value is taken, but just not enabled to edit it).
- The checked state of "Overwrite Field Format" is keeped in the config.
- Tooltip on "Overwrite Field Format" box: 
   > The field format allows to customize the format in which the date will be stored in the datasource. 
   > Overwriting the field format is usually intended for fields that are **_not_** of a date type.
   > For the display format of the date widget, use the section above.

Resolves confusion at #43177

signedav committed:
rearrange that field format is below widget display, because most users are only interested in the Widget Display settings and are confused to be able to change the field format. This might help to resolve #43177

signedav committed:
make Field Format setting a checkable group box - always unchecked, so user needs to enable before he/she makes a change there.

nyalldawson commented:
Could we avoid including "QDate" in the text? That's a qt library detail which has little/no meaning for non qt developers.

signedav committed:
don't mention QDate in tooltip

DelazJ commented:
> To make a setting, the user needs to check. After reopening the properties the group box is unchecked again.

Not sure I understand the second sentence. Does that mean that if I open the layer properties, there is no obvious way to know if I'm already overwriting the field format?

signedav commented:
> Not sure I understand the second sentence. Does that mean that if I open the layer properties, there is no obvious way to know if I'm already overwriting the field format?

It's grayed out. What means it's set but not available to enter something. When you configured a value overwriting it before, this value is written in the text edit but the text edit is disabled (the group box where it is in). 
![Screenshot from 2023-03-02 09-13-47](https://user-images.githubusercontent.com/28384354/222370062-9ffc8bff-269d-434c-9020-f50a6c10eacc.png)

To change it, you have to check the box again.

It's kind of aligned to other similar scenarios: When "Default" is chosen the format "yyyy-MM-dd" is set. It's not editable, but still it's set:
![image](https://user-images.githubusercontent.com/28384354/222370387-fc5fc233-78d6-45c9-9754-a6f0a3781d20.png)

3nids commented:
I think the config code should check if the groupbox is checked or not? (if you enable, modify, disable)

signedav commented:
There are two ways how it could behave:

**1. Always disabled**
- Open layer properties: checkbox is unchecked and groupbox is disabled.
- Check the checkbox -> groupbox becomes enabled
- change the value
- Save and close layer properties.
- **_Open layer properties again: checkbox is unchecked and groupbox is disabled (but the value is set)_**

**2. One time enable**
- Open layer properties: checkbox is unchecked and groupbox is disabled.
- Check the checkbox -> groupbox becomes enabled
- change the value
- Save and close layer properties.
- _**Open layer properties again: checkbox is checked and groupbox is enabled**_

Currently 1 is implemented. I think it's intuitive like this. But I don't have a hard opinion and I'm open to change it.

3nids commented:
v2 sounds more coherent with way QGIS works, but maybe others have a different opinion.

DelazJ commented:
My vote would also go to 2 : If I check the "overwrite...", I'd expect it to be shown to me the next time I open the dialog, with the options I set. Having a checkbox disabled means to me that options there do not apply (they may have some default values applying but these do not have particular effects on mine). It doesn't tell me that I have changes ongoing, done by myself.

With option 1, how can I say that I do not want to use that option (assuming it is ever possible)? What is the difference between we unchecking the group box, and it being disabled by closing/reopening the dialog?

Back to the issue, and assuming I understand the scope (_not used to play with date... fields_):
1. do we really need a checkbox here? Wouldn't a more descriptive label/tooltip (as you added now) be enough and clearer?
2. if the option does not concern date/time/datetime field type, because QGIS automatically detects and assigns correctly, do we need to enable the "Format field" entry for these kinds of fields?
3. If there are situations we might want to set format for date/datetime/time fields (e.g., locale-based?), then is the suggested tooltip totally correct?

signedav commented:
Ok, then. Fine for me to make it keeping its status (approach 2) or removing it. In fact I think the rearrangement having the field format below is the main improvement here...

But your concerns make sense @DelazJ. Should field format of date type be changable at all? What do you think @3nids ?

At least it needs to be changable for project with the wrong setting made manually or because of the issue fixed here #51471(that set the locale depending display format to the field format).

signedav committed:
"Widget Display" is now called "Display Format" (IMO Widget Display does not make much sense for users).

When "Overwrite Field Format" gets unchecked again, the Field Format Combobox resets back to "Default" (this to avoid that user could think the custom value is taken, but just not enabled to edit it).

The checked state of "Overwrite Field Format" is keeped in the config.

signedav commented:
I decided for behavior 2 and but keeping the possiblity to change field format of date-columns (at least to avoid breaking backwards compatibility). See "Changes" in the description.

m-kuhn commented:
I wonder if in the tooltip we should explain what it's for before explaining what it's not for

signedav commented:
> I wonder if in the tooltip we should explain what it's for before explaining what it's not for

It sais that its for fields that are not of a date type. What would you suggest?

m-kuhn commented:
The field format allows to customize the format in which the date will be stored in the datasource.

As intro to your wording.

m-kuhn commented:
The field format allows to customize the format in which the date will be stored in the datasource.

As intro to your wording.

signedav commented:
Ah now I get your point. I'll add this

signedav committed:
update tooltip to describe what the section is for

signedav committed:
Update src/ui/editorwidgets/qgsdatetimeeditconfig.ui

Co-authored-by: Matthias Kuhn <matthias@opengis.ch>

signedav commented:
Is this good to merge?
